service/src/test/java/bio/terra/workspace/app/configuration/external/controller/FolderApiControllerTest.java

```diff
@@ -0,0 +1,545 @@
+    ApiFolder updatedFolder =
+        updateFolder(
+            workspaceId,
+            secondFolder.getId(),
+            newDisplayName,
+            newDescription,
+            /*parentFolderId=*/ null,
+            /*updateParent=*/ false);
+
+    assertEquals(newDisplayName, updatedFolder.getDisplayName());
+    assertEquals(newDescription, updatedFolder.getDescription());
+    assertEquals(secondFolder.getParentFolderId(), updatedFolder.getParentFolderId());
+  }
+
+  @Test
+  public void updateFolders_updateParentTrue_folderMovedToTopLevel() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+    ApiFolder firstFolder =
+        createFolder(workspaceId, displayName, description, /*parentFolderId=*/ null);
+    ApiFolder secondFolder =
+        createFolder(
+            workspaceId, displayName, description, /*parentFolderId=*/ firstFolder.getId());
+
+    ApiFolder updatedFolder =
+        updateFolder(
+            workspaceId,
+            secondFolder.getId(),
+            "bar", // There is already a top-level foo.
+            /*newDescription=*/ null,
+            /*parentFolderId=*/ null,
+            /*updateParent=*/ true);
+
+    assertNull(updatedFolder.getParentFolderId());
+  }
+
+  @Test
+  public void updateFolder_doesNotHaveWriteAccess_throws403() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    ApiFolder folder =
+        createFolder(workspaceId, "foo", /*description=*/ null, /*parentFolderId=*/ null);
+    doThrow(new ForbiddenException("User has no write access"))
+        .when(mockSamService)
+        .checkAuthz(
+            any(AuthenticatedUserRequest.class),
+            eq(SamConstants.SamResource.WORKSPACE),
+            anyString(),
+            eq(SamConstants.SamWorkspaceAction.WRITE));
+
+    updateFolderExpectCode(
+        workspaceId,
+        folder.getId(),
+        /*newDisplayName=*/ null,
+        /*newDescription=*/ null,
+        /*parentFolderId=*/ null,
+        true,
+        HttpStatus.SC_FORBIDDEN);
+  }
+
+  @Test
+  public void updateFolders_parentFolderNotExists_throws404() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    ApiFolder folder =
+        createFolder(workspaceId, displayName, description, /*parentFolderId=*/ null);
+
+    var newDisplayName = "sofoo";
+    var newDescription = "This is a very foo folder";
+    updateFolderExpectCode(
+        workspaceId,
+        folder.getId(),
+        newDisplayName,
+        newDescription,
+        /*parentFolderId=*/ UUID.randomUUID(),
+        /*updateParent=*/ false,
+        HttpStatus.SC_NOT_FOUND);
+  }
+
+  private ApiFolder createFolder(
```

Comment:
opt description isn't needed for most of your tests. Create a second createFolder() that doesn't take in description (that calls this createFolder with blank description).